colobot-common/src/common/stringutils_case.cpp

```diff
@@ -0,0 +1,1042 @@
+    u'\u0546', u'\u0576', u'\u0546',
+    u'\u0547', u'\u0577', u'\u0547',
+    u'\u0548', u'\u0578', u'\u0548',
+    u'\u0549', u'\u0579', u'\u0549',
+    u'\u054a', u'\u057a', u'\u054a',
+    u'\u054b', u'\u057b', u'\u054b',
+    u'\u054c', u'\u057c', u'\u054c',
+    u'\u054d', u'\u057d', u'\u054d',
+    u'\u054e', u'\u057e', u'\u054e',
+    u'\u054f', u'\u057f', u'\u054f',
+    u'\u0550', u'\u0580', u'\u0550',
+    u'\u0551', u'\u0581', u'\u0551',
+    u'\u0552', u'\u0582', u'\u0552',
+    u'\u0553', u'\u0583', u'\u0553',
+    u'\u0554', u'\u0584', u'\u0554',
+    u'\u0555', u'\u0585', u'\u0555',
+    u'\u0556', u'\u0586', u'\u0556',
+    u'\u0561', u'\u0561', u'\u0531',
+    u'\u0562', u'\u0562', u'\u0532',
+    u'\u0563', u'\u0563', u'\u0533',
+    u'\u0564', u'\u0564', u'\u0534',
+    u'\u0565', u'\u0565', u'\u0535',
+    u'\u0566', u'\u0566', u'\u0536',
+    u'\u0567', u'\u0567', u'\u0537',
+    u'\u0568', u'\u0568', u'\u0538',
+    u'\u0569', u'\u0569', u'\u0539',
+    u'\u056a', u'\u056a', u'\u053a',
+    u'\u056b', u'\u056b', u'\u053b',
+    u'\u056c', u'\u056c', u'\u053c',
+    u'\u056d', u'\u056d', u'\u053d',
+    u'\u056e', u'\u056e', u'\u053e',
+    u'\u056f', u'\u056f', u'\u053f',
+    u'\u0570', u'\u0570', u'\u0540',
+    u'\u0571', u'\u0571', u'\u0541',
+    u'\u0572', u'\u0572', u'\u0542',
+    u'\u0573', u'\u0573', u'\u0543',
+    u'\u0574', u'\u0574', u'\u0544',
+    u'\u0575', u'\u0575', u'\u0545',
+    u'\u0576', u'\u0576', u'\u0546',
+    u'\u0577', u'\u0577', u'\u0547',
+    u'\u0578', u'\u0578', u'\u0548',
+    u'\u0579', u'\u0579', u'\u0549',
+    u'\u057a', u'\u057a', u'\u054a',
+    u'\u057b', u'\u057b', u'\u054b',
+    u'\u057c', u'\u057c', u'\u054c',
+    u'\u057d', u'\u057d', u'\u054d',
+    u'\u057e', u'\u057e', u'\u054e',
+    u'\u057f', u'\u057f', u'\u054f',
+    u'\u0580', u'\u0580', u'\u0550',
+    u'\u0581', u'\u0581', u'\u0551',
+    u'\u0582', u'\u0582', u'\u0552',
+    u'\u0583', u'\u0583', u'\u0553',
+    u'\u0584', u'\u0584', u'\u0554',
+    u'\u0585', u'\u0585', u'\u0555',
+    u'\u0586', u'\u0586', u'\u0556',
+};
+
+// Lookup tables
+constinit std::array<char16_t, 65536> lowercase = {};
+constinit std::array<char16_t, 65536> uppercase = {};
+
+// true if lookup tables are initialized
+constinit bool initialized = false;
+
+// Flag to make sure only one thread initializes the lookup tables
+constinit std::once_flag flag = {};
+
+// Initializes lookup tables
+void Initialize()
+{
+    for (std::size_t i = 0; i < std::size(lowercase); i++)
+    {
+        lowercase[i] = static_cast<char16_t>(i);
+        uppercase[i] = static_cast<char16_t>(i);
+    }
+
+    for (std::size_t i = 0; i < std::size(mapping); i += 3)
+    {
+        lowercase[mapping[i]] = mapping[i + 1];
+        uppercase[mapping[i]] = mapping[i + 2];
+    }
+
+    initialized = true;
+}
+
+} // namespace
+
+char32_t StrUtils::ToLower(char32_t ch)
+{
+    if (!initialized) std::call_once(flag, Initialize);
```

Comment:
No need to change the code so much, I can change `initialized` to `std::atomic_bool` or `std::atomic_flag` to fix the data race.

Reply:
What is the point of `initialized` then? `std::call_once()` should be sufficient. Is `std::atomic_bool` or `std::atomic_flag` faster?

Reply:
I checked the code generated for `std::call_once()` and while it seems relatively lightweight, a simple check of a boolean variable, even if an atomic one, should be faster because it doesn't call any functions. `ToLower()` and `ToUpper()` might be called many times for longer strings so this is potentially a significant performance boost. I can write a benchmark to verify this.